zenpacklib.py

```diff
@@ -5614,30 +5641,75 @@ def zpl_create(self):
+                    CFG = yaml.load(stream, Loader=Loader)
+
+                if diagram_type == 'yuml':
+                    print "# Classes"
+                    for cname in sorted(CFG.classes):
+                        print "[{}]".format(cname)
+
+                    print "\n# Inheritence"
+                    for cname in CFG.classes:
+                        cspec = CFG.classes[cname]
+                        for baseclass in cspec.bases:
+                            if type(baseclass) != str:
+                                baseclass = aq_base(baseclass).__name__
+                            print "[{}]^-[{}]".format(baseclass, cspec.name)
+
+                    print "\n# Containing Relationships"
+                    for crspec in CFG.class_relationships:
+                        if crspec.cardinality == '1:MC':
+                            print "[{}]++{}-{}[{}]".format(
+                                crspec.left_class, crspec.left_relname,
+                                crspec.right_relname, crspec.right_class)
+
+                    print "\n# Non-Containing Relationships"
+                    for crspec in CFG.class_relationships:
+                        if crspec.cardinality == '1:1':
+                            print "[{}]{}-.-{}[{}]".format(
+                                crspec.left_class, crspec.left_relname,
+                                crspec.right_relname, crspec.right_class)
+                        if crspec.cardinality == '1:M':
+                            print "[{}]{}-.-{}++[{}]".format(
+                                crspec.left_class, crspec.left_relname,
+                                crspec.right_relname, crspec.right_class)
+                        if crspec.cardinality == 'M:M':
+                            print "[{}]++{}-.-{}++[{}]".format(
+                                crspec.left_class, crspec.left_relname,
+                                crspec.right_relname, crspec.right_class)
+                else:
+                    LOG.error("Diagram type '%s' is not supported.", diagram_type)
             else:
-                print "Usage: %s lint <file.yaml> | py_to_yaml <zenpack name> <__init__.py> | dump_templates <zenpack_name>" % sys.argv[0]
+                print "Usage: %s lint <file.yaml> | py_to_yaml <zenpack name> | dump_templates <zenpack_name> | class_diagram [yuml] <file.yaml>" % sys.argv[0]
+
+
+        def zenpack_templatespecs(self, zenpack_name):
+            zenpack = self.dmd.ZenPackManager.packs._getOb(zenpack_name, None)
+            if zenpack is None:
+                LOG.error("Zenpack '%s' not found." % zenpack_name)
+                return
+
+            templates = {}
+            for deviceclass in [x for x in zenpack.packables() if x.meta_type == 'DeviceClass']:
+                dc_name = deviceclass.getOrganizerName()
+                templates[dc_name] = {}
+                for template in deviceclass.getAllRRDTemplates():
```

Comment:
Is getAllRRDTemplates() really what you want here? This returns templates in subclasses too which means that there are common cases where the following `templates[dc_name][template.id]` will result in values being overwritten because dc_name/template.id is not unique in the results.